infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/FsmTaskManager.java

```diff
@@ -0,0 +1,105 @@
+package org.corfudb.infrastructure.logreplication;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.infrastructure.logreplication.replication.fsm.LogReplicationEvent;
+import org.corfudb.infrastructure.logreplication.replication.fsm.LogReplicationState;
+import org.corfudb.infrastructure.logreplication.replication.fsm.LogReplicationStateType;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.IllegalTransitionException;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.LogReplicationRuntimeEvent;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.LogReplicationRuntimeState;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.LogReplicationRuntimeStateType;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * This class manages the tasks submitted by the runtime FSM and the replication FSM.
+ * Both the FSMs have their own instances of this class to which tasks are submitted and executed.
+ *
+ * The threads are shared by all the sessions traversing an FSM.
+ */
+@Slf4j
+public class FsmTaskManager {
+
+    private ExecutorService fsmWorker;
+
```

Comment:
I thought we could still have 2 (or even 4?) executors to differentiate the events from FSM and runtimeFSM, would be easier to tune the thread number and coalescing all the executors into one thread pool seems to be eliminating all the stages. 

Like in concurrent snapshot sync case, the ones finish leadership & negotiation first will be likely to have to wait for some other sessions to finish those tasks to be able to proceed

Reply:
I was thinking that the thread count can be similar for both the FSMs...

Also, I was thinking of making the sinkManager and Ack poller configurable as well, but that would mean a lot of temporary code addition in many classes. It may be a challenge to remember them. So now think that we can go with patching our code during our the tuning exercise.